### PR TITLE
added implementation to prettify error message as middleware

### LINKED ISSUE #1282: Cannot override the default error pages which contain sensitive info
**Is your feature request related to a problem? Please describe.**
The default error pages cannot be overridden and they contain sensitive info such as the internal error message and stack trace.
This is a security concern.

**Describe the solution you'd like**
* The default error pages should either have no body or shouldn't show any dangerous info such as internal error messages or stack traces in production.
* The error pages should be fully customizable. They are currently hard-coded in a few places and cannot be replaced.
* Provide an error handler to allow the user to customize the response in certain failure scenarios.

**Describe alternatives you've considered**
Overriding the error pages or mapping the `Response` object to remove the body of non 2xx pages. Neither are currently supported by `zio-http`. 

**Additional context**
Screenshot 2022-06-01 at 17 09 11

### PR DESCRIPTION
**Pull request by gciuloaica**

### CODE REVIEW

**Comment by tusharmath** [comment]
There are some places in the `Handler` also where we catch exceptions and create a 500 exception. Ideally, those too should handled by this middleware, otherwise we will have a bit of inconsistent way to display error messages.

**Comment by gciuloaica** [comment]
> There are some places in the `Handler` also where we catch exceptions and create a 500 exception. Ideally, those too should handled by this middleware, otherwise we will have a bit of inconsistent way to display error messages.

@tusharmath the error handling in `Handler` are happening after the `HttpApp` has been evaluated, so the middleware execution has been completed (unsafeRun completed). One way to unify both parts will be to migrate the formatting code to a different place (outside middleware implementation) and re-use it in Handler. To implement this it will require to know in the Handler if the middleware is in the pipeline or not.

File: zio-http/src/main/scala/zhttp/http/middleware/WebWithError.scala [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package zhttp.http.middleware
+
+class WebWithError {}
```
Comment (jdegoes) [inline zio-http/src/main/scala/zhttp/http/middleware/WebWithError.scala:3]: Dead code, let's delete

File: zio-http/src/main/scala/zhttp/http/middleware/package.scala [hunk 2]
```diff
@@ -3,4 +3,6 @@ package zhttp.http
 package object middleware {
   type HttpMiddleware[-R, +E]       = MonoMiddleware[R, E, Request, Response]
   type MonoMiddleware[-R, +E, A, B] = Middleware[R, E, A, B, A, B]
+
+  type HttpMiddlewareWithError[-R, +E] = MonoMiddleware[R, E, Option[Request], Option[Response]]
```
Comment (jdegoes) [inline zio-http/src/main/scala/zhttp/http/middleware/package.scala:7]: Dead alias, let's delete

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 3]
```diff
@@ -25,70 +25,62 @@ private[zhttp] final case class Handler[R](
...
-
-              override def unsafeContext: Ctx = ctx
-            },
-          )
-        catch {
-          case throwable: Throwable => resWriter.write(throwable, jReq)
-        }
-
-      case msg: HttpContent =>
+        unsafeRun(
```
Comment (jdegoes) [inline zio-http/src/main/scala/zhttp/service/Handler.scala:54]: These changes seem unnecessary for the scope of this PR. We should aim to do this as a "pure middleware" that does not touch anything else in the library.

**Review by jdegoes — CHANGES_REQUESTED** [review]
Would love to get this in!

Please address comments, fix conflicts, and rebase. Thank you!

**Comment by devsprint** [comment]
Closing this in favor of #1492

**Pull request closed without merging**